Structured output for 'kubeadm token list'

bart0sh opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This is an initial implementation of kubeadm structured output described in this [KEP](https://github.com/kubernetes/enhancements/pull/1054)

It adds new api group output.kubeadm.k8s.io to define versioned structured output and implements
output in TEXT, JSON, YAML, Go template and JSONPATH for ```kubeadm token list``` command using cli-runtime printer APIs. 

Example output:

```
$ kubeadm token list
TOKEN                     TTL         EXPIRES                     USAGES                   DESCRIPTION                                                EXTRA GROUPS
7vg8cr.pks5g06s84aisb27   <invalid>   2019-06-05T17:13:55+03:00   authentication,signing   The default bootstrap token generated by 'kubeadm init'.   system:bootstrappers:kubeadm:default-node-token
vkmbz7.z52etl221i9lcrvt   2h          2019-06-06T17:22:48+03:00   authentication,signing   <none>                                                     system:bootstrappers:kubeadm:default-node-token

$ kubeadm token list -o json
{
    "kind": "BootstrapToken",
    "apiVersion": "output.kubeadm.k8s.io/v1alpha1",
    "creationTimestamp": null,
    "token": "7vg8cr.pks5g06s84aisb27",
    "description": "The default bootstrap token generated by 'kubeadm init'.",
    "expires": "2019-06-05T14:13:55Z",
    "usages": [
        "authentication",
        "signing"
    ],
    "groups": [
        "system:bootstrappers:kubeadm:default-node-token"
    ]
}
{
    "kind": "BootstrapToken",
    "apiVersion": "output.kubeadm.k8s.io/v1alpha1",
    "creationTimestamp": null,
    "token": "vkmbz7.z52etl221i9lcrvt",
    "expires": "2019-06-06T14:22:48Z",
    "usages": [
        "authentication",
        "signing"
    ],
    "groups": [
        "system:bootstrappers:kubeadm:default-node-token"
    ]
}


$ kubeadm token list -o yaml
apiVersion: output.kubeadm.k8s.io/v1alpha1
creationTimestamp: null
description: The default bootstrap token generated by 'kubeadm init'.
expires: "2019-06-05T14:13:55Z"
groups:
- system:bootstrappers:kubeadm:default-node-token
kind: BootstrapToken
token: 7vg8cr.pks5g06s84aisb27
usages:
- authentication
- signing
---
apiVersion: output.kubeadm.k8s.io/v1alpha1
creationTimestamp: null
expires: "2019-06-06T14:22:48Z"
groups:
- system:bootstrappers:kubeadm:default-node-token
kind: BootstrapToken
token: vkmbz7.z52etl221i9lcrvt
usages:
- authentication
- signing
---

$ kubeadm token list -o "go-template={{println .token .description .groups}}"
7vg8cr.pks5g06s84aisb27 The default bootstrap token generated by 'kubeadm init'. [system:bootstrappers:kubeadm:default-node-token]
vkmbz7.z52etl221i9lcrvt <nil> [system:bootstrappers:kubeadm:default-node-token]

$ kubeadm token list -o "jsonpath={.token} {.description}{'\n'}}"
vg8cr.pks5g06s84aisb27 The default bootstrap token generated by 'kubeadm init'.
vkmbz7.z52etl221i9lcrvt
```

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: implemented structured output of 'kubeadm token list' in JSON, YAML, Go template and JsonPath formats
```

bart0sh committed:
create new api group output.kubeadm.k8s.io

This group contains APIs for handling kubeadm
structured output.

bart0sh committed:
kubeadm token list: implement structured output

Used cli-runtime API to print bootstrap tokens in 5 formats:

 - TEXT (identical to the current output)
 - YAML
 - JSON
 - JSONPATH
 - Go template

bart0sh commented:
/cc @akutz

bart0sh commented:
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

bart0sh commented:
/cc @rosti @timothysc @neolit123

bart0sh commented:
/test pull-kubernetes-bazel-test

bart0sh commented:
/test pull-kubernetes-e2e-gce

kad commented:
lgtm from my side, however question: should  the output of bootstrap token in parsable form to be compatible with something that can be used directly in joinconfig?

rosti commented on cmd/kubeadm/app/cmd/options/generic.go:
```diff
@@ -80,3 +82,10 @@ func AddKubernetesVersionFlag(fs *pflag.FlagSet, kubernetesVersion *string) {
+func AddOutputFlag(fs *pflag.FlagSet, outputFormat *string) {
+	fs.StringVarP(
+		outputFormat, "output", "o", output.TEXTOutput,
+		fmt.Sprintf("Output format; available options are: %s, %s, %s and Go template", output.TEXTOutput, output.YAMLOutput, output.JSONOutput))
```
We might have a slice with all supported output formats for this purpose and for validation too.

rosti commented on cmd/kubeadm/app/cmd/options/generic.go:
```diff
@@ -80,3 +82,10 @@ func AddKubernetesVersionFlag(fs *pflag.FlagSet, kubernetesVersion *string) {
+// AddOutputFlag adds the --output/-o flag to the given flagset
+func AddOutputFlag(fs *pflag.FlagSet, outputFormat *string) {
+	fs.StringVarP(
+		outputFormat, "output", "o", output.TEXTOutput,
```
Might start with a `--experimental-output`. Also, not sure if `o` would be available for an alpha feature. Might want to just reserve it for now (but this needs discussion).

bart0sh replied:
I thought that it's safe to add new option as we're not modifying existing option here. It's not a big deal to change though. Agree, we need to discuss this.

fabriziopandini replied:
+1 for `--experimental-output` and `-o`

bart0sh replied:
done

rosti reviewed (commented):
Thanks @bart0sh !
Looks good for a start. I propose to reuse the corresponding types from the config (when possible).

bart0sh commented on cmd/kubeadm/app/apis/output/types.go:
```diff
@@ -0,0 +1,51 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:",inline"`
+
+	// Token is used for establishing bidirectional trust between nodes and control-planes.
```
@rosti thank you for the review. I tried that. It broke boilerplate generation this way:
```
$ make generated_files
E0611 17:02:00.288082   26597 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/cmd/kubeadm/app/apis/output/v1alpha1.BootstrapToken -> k8s.io/kubernetes/cmd/kubeadm/app/apis/output.BootstrapToken
E0611 17:02:00.288293   26597 conversion.go:668]   the following fields need manual conversion:
E0611 17:02:00.288299   26597 conversion.go:670]       - BootstrapToken
E0611 17:02:00.288419   26597 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/cmd/kubeadm/app/apis/output.BootstrapToken -> k8s.io/kubernetes/cmd/kubeadm/app/apis/output/v1alpha1.BootstrapToken
E0611 17:02:00.288425   26597 conversion.go:668]   the following fields need manual conversion:
E0611 17:02:00.288428   26597 conversion.go:670]       - Token
E0611 17:02:00.288432   26597 conversion.go:670]       - Description
E0611 17:02:00.288437   26597 conversion.go:670]       - TTL
E0611 17:02:00.288442   26597 conversion.go:670]       - Expires
E0611 17:02:00.288446   26597 conversion.go:670]       - Usages
E0611 17:02:00.288450   26597 conversion.go:670]       - Groups
```

I didn't want to add manual conversion as it'd mean more maintenance effort. However, I agree with you. This duplication doesn't look good.

bart0sh replied:
@rosti fixed this one eventually.

bart0sh commented on cmd/kubeadm/app/apis/output/types.go:
```diff
@@ -0,0 +1,51 @@
+// metav1.TypeMeta and metav1.ObjectMeta
+type BootstrapToken struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:",inline"`
```
thanks, will remove those

bart0sh commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -0,0 +1,51 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:",inline"`
+
+	// Token is used for establishing bidirectional trust between nodes and control-planes.
```
see above

bart0sh commented:
@kad thank you for the review. Agree, it makes sense to reuse existing token related structures. However, this would require extending existing structures, which I was hesitating to do in this PR.
We need to discuss this further I guess.

bart0sh commented:
/test pull-kubernetes-e2e-gce

bart0sh commented:
/test pull-kubernetes-integration

xmudrii commented:
@bart0sh Hello! I'm bug triage lead for the 1.16 release cycle and considering this PR has not been updated for a long time, I'd like to check what's the status of this PR. The code freeze is starting 29th August (about 2.5 weeks from now) and while there is plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.16, is it still planned for this release?

neolit123 commented:
> As the PR is tagged for 1.16, is it still planned for this release?

hi @xmudrii , no it is not.
i will clear the milestone.

/milestone clear

bart0sh committed:
Test token output

Tested JSON, YAML, Go Template and Text token output formats.

bart0sh committed:
Add internal kubeadm output API

Added internal structures and APIs to handle
kubeadm structured output.

bart0sh commented:
@kubernetes/sig-cluster-lifecycle-pr-reviews 

Updated this PR:
- added tests
- used external cli-runtime API
- added structured output for 'kubeadm upgrade plan'

please, review again

bart0sh commented:
@rosti @neolit123 @fabriziopandini please, review and let me know if this direction is ok. Thanks.

fabriziopandini commented on cmd/kubeadm/app/apis/output/v1alpha1/defaults.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+// SetDefaults_BootstrapToken assigns default values to the BootstrapToken
+func SetDefaults_BootstrapToken(bt *BootstrapToken) {
```
I don't think the structured output should do any defaulting at all because, as far as understand, it is responsible only for the output format and it should not alter data

bart0sh replied:
done

fabriziopandini commented on cmd/kubeadm/app/apis/output/doc.go:
```diff
@@ -0,0 +1,21 @@
+// +k8s:deepcopy-gen=package
+// +groupName=output.kubeadm.k8s.io
+
+// Package output implements the kubeadm structured output
```
[1] It will be great to one or two sentences  about why we are adding this and which are the expected consumers (probably in the KEP there is already something we can copy/paste)

bart0sh replied:
Done

fabriziopandini commented on cmd/kubeadm/app/apis/output/v1alpha1/doc.go:
```diff
@@ -0,0 +1,24 @@
+// +k8s:deepcopy-gen=package
+// +k8s:conversion-gen=k8s.io/kubernetes/cmd/kubeadm/app/apis/output
+
+// Package v1alpha1 defines the v1alpha1 version of the kubeadm data structures
```
[1]

bart0sh replied:
Done

fabriziopandini commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -0,0 +1,63 @@
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BootstrapToken represents information for the output produced by 'kubeadm token list'
```
Even if the scope of the current PR is 'kubeadm token list', I think that the same output can be used e.g. for `kubeadm token create` or eventually also `kubeadm init`.  So I will make this description more generic saying this represents the output form kubeadm command returning or generating Bootstrap tokens. WDYT?

bart0sh replied:
done

fabriziopandini commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -0,0 +1,63 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:",inline"`
+
+	Component      string `json:"name,omitempty"`
```
what about adding an enum?

bart0sh replied:
will do in the 'upgrade plan' PR

fabriziopandini commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -0,0 +1,63 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:",inline"`
+
+	Component      string `json:"name,omitempty"`
```
Remove `,omitempty`? (same for `currentVersion`, `newVersion`)

bart0sh replied:
done

fabriziopandini reviewed (changes requested):
@bart0sh great to see this moving! 
IMO we are definitely on the right path; my two main suggestion to get this moving faster are:
- Consider if to split the `upgrade plan` implementation in a separated PR, so this PR will reduce it size and it will be easier to get it merged.
- Consider if to move printers implementation from `cmd` files to a separated `output` package, so cmd will be simplified, printers implementation will be isolated, and printers might be reused across commands

Few other minor suggestions are inline

bart0sh commented:
@fabriziopandini Thank you for the review! I've update the PR. Please, review again.

bart0sh commented:
> Consider if to split the upgrade plan implementation in a separated PR, so this PR will reduce it size and it will be easier to get it merged.

The point of including it into this PR is to show the way to do it for 2 different type of commands: token list and upgrade plan. I can do it in a separate PR, but it will depend on this one and will fail all possible checks.
So, I'm going to submit it as a separate PR after this PR is merged.

bart0sh commented:
/test pull-kubernetes-integration
/test pull-kubernetes-bazel-test
/test pull-kubernetes-kubemark-e2e-gce-big

bart0sh commented:
/test pull-kubernetes-bazel-test
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-verify

bart0sh commented:
/test pull-kubernetes-bazel-test

bart0sh commented:
/test pull-kubernetes-node-e2e

bart0sh commented:
/test pull-kubernetes-node-e2e

bart0sh commented:
@fabriziopandini @neolit123 @rosti I've addressed pretty much all your comments. Please, review again and merge if you don't mind.

Klaven reviewed (changes requested):
This is looking really awesome. I pulled the code, ran it, ran tests with it and am quite impressed! Thanks for the hard work making this a reality!

I have one issue mostly from a usability standpoint. but other then that I think this is a good start on machine readable output! thank you.

bart0sh commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -155,9 +162,11 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 		},
 	}
+
+	options.AddOutputFlag(listCmd.Flags(), &outputFormat)
```
@Klaven thank you for the review!

Your comment convinced me to consider using [genericoptions API](https://github.com/kubernetes/cli-runtime/tree/master/pkg/genericclioptions), which has better validation options and supports extra output formats (gotemplate-file, jsonpath, jsonpath-file):

```
$ kubeadm token list -o bla
unable to match a printer suitable for the output format "bla", allowed formats are: go-template,go-template-file,json,jsonpath,jsonpath-file,template,templatefile,text,yaml
To see the stack trace of this error execute with --v=5 or higher
```

```
$ kubeadm token list --help | grep ^Flags -A2
Flags:
      --allow-missing-template-keys   If true, ignore any errors in templates when a field or map key is missing in the template. Only applies to golang and jsonpath output formats. (default true)
  -o, --experimental-output string    Output format. One of: text|json|yaml|go-template|go-template-file|template|templatefile|jsonpath|jsonpath-file. (default "text")
```

bart0sh commented:
@fabriziopandini @neolit123 @rosti @Klaven 

Latest update: used [genericoptions](https://github.com/kubernetes/cli-runtime/tree/master/pkg/genericclioptions) API from cli-runtime, which has better validation options and supports additional  output formats (gotemplate-file, jsonpath, jsonpath-file)

Please, review again and merge if you don't mind.

neolit123 commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -268,10 +283,72 @@ func RunGenerateToken(out io.Writer) error {
+type tokenTextPrintFlags struct{}
+
+// ToPrinter returns resource printer for the text output format
+func (tpf tokenTextPrintFlags) ToPrinter(outputFormat string) (output.KubeadmPrinter, error) {
```
unless there is demand to copy `tpf` in memory on each call to `ToPrinter`, this should be:
`func (tpf *tokenTextPrintFlags)`.

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// NewKubeadmOutputFlags creates new KubeadmOutputFlags
+func NewKubeadmOutputFlags(kubeadmTextPrintFlags KubeadmTextPrintFlags) *KubeadmOutputFlags {
```
this should accept a pointer to `KubeadmTextPrintFlags`.

bart0sh replied:
done

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,201 @@
+
+	TypeSetterPrinter *printers.TypeSetterPrinter
+
+	OutputFormat *string
```
please add descriptions above all the fields of this structure.

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,201 @@
+// AllowedFormats returns list of allowed output formats
+func (kof *KubeadmOutputFlags) AllowedFormats() []string {
+	ret := []string{"text"}
+
```
empty line can be omitted.

bart0sh replied:
done

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// PrintObj is an implementation of ResourcePrinter.PrintObj that prints object
+func (tp TextPrinter) PrintObj(obj runtime.Object, writer io.Writer) error {
```
better have the `tp *TextPrinter` signature for all methods.

bart0sh replied:
done

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,201 @@
+func (kof *KubeadmOutputFlags) AddFlags(cmd *cobra.Command) {
+	kof.JSONYamlPrintFlags.AddFlags(cmd)
+	kof.KubeTemplatePrintFlags.AddFlags(cmd)
+	cmd.Flags().StringVarP(kof.OutputFormat, "experimental-output", "o", *kof.OutputFormat, fmt.Sprintf("Output format. One of: %s.", strings.Join(kof.AllowedFormats(), "|")))
```
question for everyone: given the long command contains experimental, do we want the shortcut to be just `o`?

bart0sh replied:
it was discussed earlier in this PR. see above.

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// PrintObj is an implementation of ResourcePrinter.PrintObj that calls underlying printer API
+func (rpw ResourcePrinterWrapper) PrintObj(obj runtime.Object, writer io.Writer) error {
```
[1] same for this.

bart0sh replied:
done

neolit123 commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -0,0 +1,34 @@
+// metav1.TypeMeta and metav1.ObjectMeta
+type BootstrapToken struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:",inline"`
```
@rosti you had objections about adding the full metav1 types to our objects, because we don't need a lot of the automatic/dynamic fields that the api server manages.

yet, the structured output needs them.
maybe we should still include the full ObjectMeta there for consistency instead of defining a subset.

@bart0sh 
i remember we were in doubt whether we should duplicate BoostrapToken under the output group.
we may still wan to do that, but wondering whether 
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go#L228
should have TypeMeta and ObjectMeta
and this structure should look like:

```
type BootstrapToken struct {
    kubeadmapiv1beta2.BootstrapToken
}
```

bart0sh replied:
I'd prefer not to modify kubeadm API just for this. Another point is that  kubeadmapiv1beta2.BootstrapToken doesn't need TypeMeta and ObjectMeta fields as they're only required by output machinery, hence it's better to have them only in the output API in my opinion.

neolit123 commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -268,10 +283,72 @@ func RunGenerateToken(out io.Writer) error {
+	return fmt.Sprintf("%s\t%s\t%s\t%s\t%s\t%s\n", obj.Token, ttl, expires, usages, description, groups)
+}
+
+// upgradePlanPrinter prints upgrade plan in a text form
```
change comment to relate to `tokenTextPrinter`

bart0sh replied:
done

neolit123 commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+// Info is a wrapper around klog.Info
+func (tp TextPrinter) Info(level klog.Level, args ...interface{}) {
```
i think we should either only have Infof or both Info and Infof.
also we need Warning(f).

this however raises another question.
users would expected structured output to go to STDOUT.
klog goes to STDERR, so do we need to wrap it with the structured output printer?

bart0sh replied:
added Infof.

structured output is handled only by PrintObj. Info* calls area silenced in the current implementation. I'm ok to make them loud if other reviewers agree. It would create a bit of messy console output though.

neolit123 commented:
@bart0sh 
thanks a lot for the work on this. i have added a number of comments in my first pass.
i think this PR is in a good shape but needs more review from others.
@kubernetes/sig-cluster-lifecycle-pr-reviews 

i will do another pass myself at some point.

bart0sh commented:
/test pull-kubernetes-kubemark-e2e-gce-big

rosti commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,216 @@
+// KubeadmPrinter is a common printing interface in Kubeadm
+type KubeadmPrinter interface {
+	PrintObj(obj runtime.Object, writer io.Writer) error
+	Info(level klog.Level, args ...interface{})
```
Structured output should go to `os.Stdout`, while logs should be on `os.Stderr`. Hence, `Info` and `Infof` are not required.

bart0sh replied:
Done. i don't fully agree with this as it would mess up UX and confuse users. Human beings are not able to distinguish between stdout and stderr output when they look at the console :)

neolit123 replied:
maybe we should just document it?
handling all the klog methods is probably not something we want to do.

but also the feature is alpha, so based on feedback we can change that.

bart0sh replied:
Sure, let's merge it and see what people tell us.

neolit123 replied:
waiting on the KEP approval, only @timothysc can do that.

rosti reviewed (commented):
Thanks @bart0sh !
Looks good for a start. Mostly naming stuff that seems to need fixing and logging.

bart0sh commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -268,10 +283,73 @@ func RunGenerateToken(out io.Writer) error {
+func RunListTokens(out io.Writer, errW io.Writer, client clientset.Interface, printer output.KubeadmPrinter) error {
 	// First, build our selector for bootstrap tokens only
-	klog.V(1).Infoln("[token] preparing selector for bootstrap token")
+	printer.Info(1, "[token] preparing selector for bootstrap token")
```
@rosti Thank you very much for the review!

UX will suffer in some cases I believe as info output will be mixed with structure output, but if you're ok with this I'm ok too. Returned back klog usage.

bart0sh commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,216 @@
+)
+
+// KubeadmTextPrintFlags is an iterface to handle custom text output
+type KubeadmTextPrintFlags interface {
```
Done

bart0sh commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,216 @@
+// KubeadmOutputFlags composes common printer flag structs
+// used across kubeadm commands, and provides a method
+// of retrieving a known printer based on flag values provided.
+type KubeadmOutputFlags struct {
```
*Flags suffix is used in the [cli-runtime code](https://github.com/kubernetes/cli-runtime/blob/master/pkg/genericclioptions/json_yaml_flags.go#L37). I'd rather use their naming convention.

P.S.: Would s/OutputFlags/OutputPrintFlags/ go? cli runtime uses  PrintFlags suffix, so this way it would be even closer to their names.

neolit123 replied:
but the structure seems to contain more than just flags?

bart0sh replied:
Yes, it contains flags and additional fields to serve its second purpose that's mentioned in the comment - to retrive a printer based on the flag values.

bart0sh commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,216 @@
+	KubeTemplatePrintFlags *genericclioptions.KubeTemplatePrintFlags
+	// JSONYamlPrintFlags provides default flags necessary for kubeadm text printing.
+	KubeadmTextPrintFlags KubeadmTextPrintFlags
+	//TypeSetterPrinter is an implementation of ResourcePrinter wraps another printer with types set on the objects
```
Done

bart0sh commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,216 @@
+
+// AllowedFormats returns list of allowed output formats
+func (kof *KubeadmOutputFlags) AllowedFormats() []string {
+	ret := []string{"text"}
```
Done

bart0sh commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,216 @@
+
+	if kof.JSONYamlPrintFlags != nil {
+		if p, err := kof.JSONYamlPrintFlags.ToPrinter(outputFormat); !genericclioptions.IsNoCompatiblePrinterError(err) {
+			return NewResourcePrinterWrapper(kof.TypeSetterPrinter.WrapToPrinter(p, err))
```
This approach was also taken from the [cli-runtime code](https://github.com/kubernetes/cli-runtime/blob/master/pkg/genericclioptions/print_flags.go#L84:L119) I used it this way just to not reinvent the wheel and be consistent with their implementation.

Let me know if you insist on using helpers - I'll modify the code.

bart0sh commented on cmd/kubeadm/app/util/output/output.go:
```diff
@@ -0,0 +1,216 @@
+}
+
+// KubeadmPrinter is a common printing interface in Kubeadm
+type KubeadmPrinter interface {
```
Done

bart0sh commented:
/test pull-kubernetes-kubemark-e2e-gce-big

fabriziopandini commented:
@bart0sh Thanks for your great job on this PR! 
/approve
Leaving final LGTM to some other reviewer involved in this loop

bart0sh commented:
@fabriziopandini Thank you very much for approving this PR! There will be more of this kind as soon as this one is merged.

neolit123 commented:
voting on final merge on this PR during today's office hours.
seems good to go, we can amend later if needed.

neolit123 commented:
/lgtm
